[PRW 2.0]: Migrate gogo/proto to vtproto

tonyli233 opened the pull request:
<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->
*This work is under [Google Summer of Code 2024](https://summerofcode.withgoogle.com/programs/2024/projects/s9tV51nw). The last commit before the submission is https://github.com/prometheus/prometheus/pull/14904/commits/7ec814a0b27d31dcb2ae03174077fa7b7499ddeb*.

This PR migrates the RW2.0 from [gogo/proto](https://github.com/gogo/protobuf) to [vtproto](https://github.com/planetscale/vtprotobuf/tree/main). The performance and memory usage of vtproto compared to gogo/proto were worried by the community. This proposal proves the possibility of migration while not damaging the performance and memory usage. The pooling of vtproto is enabled for all types in `io.prometheus.write.v2.types`.

We gave an effort for [hand-written marshaling](https://github.com/prometheus/prometheus/pull/14394). This proposal was given up since the hand-written marshal functions are quite complicated and hard to maintain. 

- The benchmark result for sending samples
```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/storage/remote
cpu: Apple M1
                                            │ ../main_bench.txt │          ../vt20_bench.txt           │
                                            │      sec/op       │    sec/op     vs base                │
SampleSend/prometheus.WriteRequest-8               4.671m ± 37%   4.460m ± 11%        ~ (p=0.684 n=10)
SampleSend/io.prometheus.write.v2.Request-8        2.579m ±  2%   1.197m ±  5%  -53.57% (p=0.000 n=10)
StoreSeries/plain-8                               1452.9µ ±  0%   903.2µ ±  0%  -37.83% (p=0.000 n=10)
StoreSeries/externalLabels-8                       2.349m ±  0%   1.478m ±  0%  -37.06% (p=0.000 n=10)
StoreSeries/relabel-8                              1.917m ±  1%   1.212m ±  1%  -36.76% (p=0.000 n=10)
StoreSeries/externalLabels+relabel-8               2.570m ±  0%   1.623m ±  1%  -36.85% (p=0.000 n=10)
BuildWriteRequest/2_instances-8                    23.57µ ±  0%   25.08µ ±  0%   +6.42% (p=0.000 n=10)
BuildWriteRequest/10_instances-8                   114.1µ ±  4%   124.9µ ±  0%   +9.50% (p=0.002 n=10)
BuildWriteRequest/1k_instances-8                   1.140m ±  0%   1.241m ±  0%   +8.88% (p=0.000 n=10)
BuildV2WriteRequest/2_instances-8                  29.62µ ±  1%   17.99µ ±  2%  -39.24% (p=0.000 n=10)
BuildV2WriteRequest/10_instances-8                144.74µ ±  2%   85.56µ ±  2%  -40.88% (p=0.000 n=10)
BuildV2WriteRequest/1k_instances-8                1310.8µ ±  6%   821.4µ ±  1%  -37.33% (p=0.000 n=10)
BuildTimeSeries-8                                  2.486m ±  3%   2.325m ±  3%   -6.50% (p=0.000 n=10)
StreamReadEndpoint-8                               64.02µ ± 54%   21.44µ ±  3%  -66.52% (p=0.000 n=10)
RemoteWriteHandler-8                               5.084µ ± 16%   4.194µ ±  2%  -17.51% (p=0.000 n=10)
geomean                                            426.6µ         300.1µ        -29.66%

                                            │ ../main_bench.txt │              ../vt20_bench.txt              │
                                            │       B/op        │      B/op       vs base                     │
SampleSend/prometheus.WriteRequest-8              848.2Ki ± 33%    589.8Ki ± 32%   -30.46% (p=0.005 n=10)
SampleSend/io.prometheus.write.v2.Request-8       81.21Ki ± 10%    58.05Ki ±  2%   -28.52% (p=0.000 n=10)
StoreSeries/plain-8                               406.0Ki ±  0%    406.0Ki ±  0%         ~ (p=0.494 n=10)
StoreSeries/externalLabels-8                      1.007Mi ±  0%    1.007Mi ±  0%    -0.00% (p=0.028 n=10)
StoreSeries/relabel-8                             906.3Ki ±  0%    906.3Ki ±  0%         ~ (p=0.698 n=10)
StoreSeries/externalLabels+relabel-8             1000.3Ki ±  0%   1000.2Ki ±  0%    -0.00% (p=0.023 n=10)
BuildWriteRequest/2_instances-8                    82.000 ±  0%      2.000 ±  0%   -97.56% (p=0.000 n=10)
BuildWriteRequest/10_instances-8                   146.00 ±  1%      54.50 ±  3%   -62.67% (p=0.000 n=10)
BuildWriteRequest/1k_instances-8                  6.146Ki ±  0%    5.205Ki ±  0%   -15.31% (p=0.000 n=10)
BuildV2WriteRequest/2_instances-8                   1.000 ±  0%      0.000 ±  0%  -100.00% (p=0.000 n=10)
BuildV2WriteRequest/10_instances-8                  24.00 ±  8%      13.00 ±  8%   -45.83% (p=0.000 n=10)
BuildV2WriteRequest/1k_instances-8                1.983Ki ±  7%    1.222Ki ±  1%   -38.39% (p=0.000 n=10)
BuildTimeSeries-8                                 2.527Mi ±  0%    3.210Mi ±  0%   +27.02% (p=0.000 n=10)
StreamReadEndpoint-8                              23.66Ki ±  0%    23.47Ki ±  0%    -0.82% (p=0.000 n=10)
RemoteWriteHandler-8                              3.258Ki ±  0%    3.269Ki ±  0%    +0.33% (p=0.000 n=10)
geomean                                           12.06Ki                         ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean

                                            │ ../main_bench.txt │             ../vt20_bench.txt             │
                                            │     allocs/op     │  allocs/op    vs base                     │
SampleSend/prometheus.WriteRequest-8             1.133k ± 31%     3.776k ± 32%  +233.13% (p=0.000 n=10)
SampleSend/io.prometheus.write.v2.Request-8       605.5 ±  3%      509.0 ±  4%   -15.94% (p=0.000 n=10)
StoreSeries/plain-8                              1.604k ±  0%     1.604k ±  0%         ~ (p=0.474 n=10)
StoreSeries/externalLabels-8                     2.608k ±  0%     2.608k ±  0%         ~ (p=1.000 n=10)
StoreSeries/relabel-8                            5.603k ±  0%     5.603k ±  0%         ~ (p=1.000 n=10)
StoreSeries/externalLabels+relabel-8             5.607k ±  0%     5.607k ±  0%         ~ (p=0.474 n=10)
BuildWriteRequest/2_instances-8                   1.000 ±  0%      0.000 ±  0%  -100.00% (p=0.000 n=10)
BuildWriteRequest/10_instances-8                  1.000 ±  0%      0.000 ±  0%  -100.00% (p=0.000 n=10)
BuildWriteRequest/1k_instances-8                  20.00 ±  0%      57.00 ±  0%  +185.00% (p=0.000 n=10)
BuildV2WriteRequest/2_instances-8                 0.000 ±  0%      0.000 ±  0%         ~ (p=1.000 n=10) ¹
BuildV2WriteRequest/10_instances-8                0.000 ±  0%      0.000 ±  0%         ~ (p=1.000 n=10) ¹
BuildV2WriteRequest/1k_instances-8                21.00 ± 10%      20.00 ±  0%    -4.76% (p=0.000 n=10)
BuildTimeSeries-8                                39.75k ±  0%     69.75k ±  0%   +75.48% (p=0.000 n=10)
StreamReadEndpoint-8                              234.0 ±  0%      273.0 ±  0%   +16.67% (p=0.000 n=10)
RemoteWriteHandler-8                              38.00 ±  0%      43.00 ±  0%   +13.16% (p=0.000 n=10)
geomean                                                       ²                 ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

                                   │ ../main_bench.txt │             ../vt20_bench.txt              │
                                   │ compressedSize/op │ compressedSize/op  vs base                 │
BuildWriteRequest/2_instances-8            1.933k ± 0%         1.933k ± 0%       ~ (p=1.000 n=10) ¹
BuildWriteRequest/10_instances-8           8.156k ± 0%         8.156k ± 0%       ~ (p=1.000 n=10) ¹
BuildWriteRequest/1k_instances-8           78.22k ± 0%         78.22k ± 0%       ~ (p=1.000 n=10) ¹
BuildV2WriteRequest/2_instances-8          1.517k ± 0%         1.493k ± 0%  -1.58% (p=0.000 n=10)
BuildV2WriteRequest/10_instances-8         5.338k ± 0%         5.148k ± 0%  -3.56% (p=0.000 n=10)
BuildV2WriteRequest/1k_instances-8         48.80k ± 0%         46.69k ± 0%  -4.33% (p=0.000 n=10)
geomean                                    8.871k              8.729k       -1.59%
¹ all samples are equal
```
We see no significant change in the runtime while around a 46% increase in memory usage. This proves the possibility of using vtproto with pooling for v2 while we still need to support v1 using similar method. 

- The memory allocation of gogo/proto for a Prometheus instance that scrapes itself while writing remotely. (v2)
<img width="1435" alt="memory_allocation_gogo" src="https://github.com/user-attachments/assets/aff53503-1303-4e65-8712-f67dc66bf363">
- The memory allocation of vtproto for a Prometheus instance that scrapes itself while writing remotely. (v2)
<img width="1440" alt="memory_allocation_vt" src="https://github.com/user-attachments/assets/fff176a8-793e-466d-bf92-ce328dcad1bf">

- The memory allocation of gogo/proto for a Prometheus instance that scrapes itself while writing remotely. (v1)
<img width="1920" alt="memory_allocation_gogo_v1" src="https://github.com/user-attachments/assets/50000e23-5311-4a7a-8758-1193b3a47003">
- The memory allocation of vtproto for a Prometheus instance that scrapes itself while writing remotely. (v1)
<img width="1919" alt="memory_allocation_vt_v1" src="https://github.com/user-attachments/assets/4a1ac175-9bfa-4bfb-a65c-4521c24aecdc"> 

From the two screenshots of the Prometheus dashboard, we can observe that the memory allocation for vtproto is less than gogo/proto. This is probably due to the vt pool. 

- The flame graph of a Prometheus receive using gogoproto (v2)
<img width="1440" alt="gogo-unmarshal" src="https://github.com/user-attachments/assets/82f89689-87e9-40d6-810c-764046db5c15">

- The flame graph of a Prometheus receiver using vtproto (v2)
<img width="1440" alt="vt-unmarshal" src="https://github.com/user-attachments/assets/9fce11a6-ac91-45c7-bed8-ae1d2b47fc52">

- The flame graph of a Prometheus receive using gogoproto (v1)
<img width="1920" alt="gogo-unmarshal-v1" src="https://github.com/user-attachments/assets/fde3e599-7e7e-493d-b99b-fe2894d26171">

- The flame graph of a Prometheus receiver using vtproto (v1)
<img width="1917" alt="vt-unmarshal-v1" src="https://github.com/user-attachments/assets/2b177e7a-03d2-4aee-9523-b4eb1ddb0881">

From the two screenshots of the flame graph, we can observe that marshaling of vtproto is faster than gogo/proto.

Tests to fix:
- `protobufparse_test.go`: This test is testing on the parsing of time series using `UnmarshalText` which is provided by gogo/proto. We need to discuss how we fix this test. 
* `scrape_test.go`: This test also uses `UnmarshalText` from gogo/proto.
* `TestPrometheusConverter_addSumNumberDataPoints` in `storage/remote/otlptranslator/prometheusremotewrite.go`: This test requires further investigation. 
* buf.build/lint: The buf lint action is checking against the main branch. This error can be ignored before merging. It should be resolved by itself after merged to main. The test fails since breaking changes were made in the pr:
  * deleted "gogoproto/gogo.proto" since we don't need this dependency
  * deleted "io/prometheus/client/metrics.proto" since this is imported from [client_model](https://github.com/prometheus/client_model).
  * deleted "google/protobuf/timestamp.proto" and "google/protobuf/descriptor.proto" since they were used in "io/prometheus/client/metrics.proto"


Todo:
- After [`planetscale/vtprotobuf`](https://github.com/bufbuild/modules/issues/692) is added to the Buf Schema Registry, add protobuf extended message option `option (vtproto.mempool) = true;` for all types in `types.proto`. We are not able to do this currently since the module is not on BSR yet. Currently, the pooling methods are generated by setting options for vtproto `buf.gen.yaml`. We can remove the list of options after we can set the message option. We can refer to the details at [vtproto pooling options](https://github.com/planetscale/vtprotobuf/tree/main#usage).

cstyan commented:
Tony ( @tonyli233 ) is mentioning that some of the benchmark results are concerning/could still be improved, but it's worth noting that some bechmarks haven't been updated to use the new v2.0 format yet and others don['t work with v2 proto (such as BenchmarkSampleSend, which the v2 version never completes atm).

His benchmark results at the top of this PR are showing gogoproto v2 vs vtproto v2, but if we compare both to main with v1 we're still seeing good results. For example:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/storage/remote
cpu: AMD Ryzen 9 5950X 16-Core Processor
              │    main     │               main-v2               │                vt20                 │
              │   sec/op    │   sec/op     vs base                │   sec/op     vs base                │
SampleSend-32   3.480m ± 9%   1.397m ± 6%  -59.86% (p=0.000 n=10)   1.356m ± 5%  -61.03% (p=0.000 n=10)

              │     main      │               main-v2                │                 vt20                 │
              │     B/op      │     B/op      vs base                │     B/op      vs base                │
SampleSend-32   648.03Ki ± 6%   55.75Ki ± 7%  -91.40% (p=0.000 n=10)   54.55Ki ± 7%  -91.58% (p=0.000 n=10)

              │    main    │              main-v2               │                vt20                │
              │ allocs/op  │ allocs/op   vs base                │ allocs/op   vs base                │
SampleSend-32   887.0 ± 5%   425.5 ± 6%  -52.03% (p=0.000 n=10)   585.0 ± 7%  -34.05% (p=0.000 n=10)
```

tonyli233 commented:
> Tony ( @tonyli233 ) is mentioning that some of the benchmark results are concerning/could still be improved, but it's worth noting that some bechmarks haven't been updated to use the new v2.0 format yet and others don['t work with v2 proto (such as BenchmarkSampleSend, which the v2 version never completes atm).
> 
> His benchmark results at the top of this PR are showing gogoproto v2 vs vtproto v2, but if we compare both to main with v1 we're still seeing good results. For example:
> 
> ```
> goos: linux
> goarch: amd64
> pkg: github.com/prometheus/prometheus/storage/remote
> cpu: AMD Ryzen 9 5950X 16-Core Processor
>               │    main     │               main-v2               │                vt20                 │
>               │   sec/op    │   sec/op     vs base                │   sec/op     vs base                │
> SampleSend-32   3.480m ± 9%   1.397m ± 6%  -59.86% (p=0.000 n=10)   1.356m ± 5%  -61.03% (p=0.000 n=10)
> 
>               │     main      │               main-v2                │                 vt20                 │
>               │     B/op      │     B/op      vs base                │     B/op      vs base                │
> SampleSend-32   648.03Ki ± 6%   55.75Ki ± 7%  -91.40% (p=0.000 n=10)   54.55Ki ± 7%  -91.58% (p=0.000 n=10)
> 
>               │    main    │              main-v2               │                vt20                │
>               │ allocs/op  │ allocs/op   vs base                │ allocs/op   vs base                │
> SampleSend-32   887.0 ± 5%   425.5 ± 6%  -52.03% (p=0.000 n=10)   585.0 ± 7%  -34.05% (p=0.000 n=10)
> ```

Thanks, @cstyan. I saw the fix to [BuildWriteRequest](https://github.com/prometheus/prometheus/pull/14925) and updated the result for all benchmarks in the `storage/remote` package here. I have also added the result to v1 using vtproto with pooling. I think the result proves that we can use vtproto.

tonyli233 commented on prompb/codec.go:
```diff
@@ -46,13 +46,13 @@ func labelProtosToLabels(b *labels.ScratchBuilder, labelPairs []Label) labels.La
-			Name:  l.Name,
-			Value: l.Value,
-		})
+		lbl := LabelFromVTPool()
+		lbl.Name = l.Name
+		lbl.Value = l.Value
+		result = append(result, lbl)
```
This is getting the `Label` type from the memory pool.

tonyli233 commented on prompb/codec.go:
```diff
@@ -146,43 +146,46 @@ func deltasToCounts(deltas []int64) []float64 {
-		Timestamp:      timestamp,
-	}
+func FromIntHistogram(timestamp int64, h *histogram.Histogram) *Histogram {
+	hist := HistogramFromVTPool()
+	hist.Count = &Histogram_CountInt{CountInt: h.Count}
+	hist.Sum = h.Sum
+	hist.Schema = h.Schema
+	hist.ZeroThreshold = h.ZeroThreshold
+	hist.ZeroCount = &Histogram_ZeroCountInt{ZeroCountInt: h.ZeroCount}
+	hist.NegativeSpans = spansToSpansProto(h.NegativeSpans)
+	hist.NegativeDeltas = h.NegativeBuckets
+	hist.PositiveSpans = spansToSpansProto(h.PositiveSpans)
+	hist.PositiveDeltas = h.PositiveBuckets
+	hist.ResetHint = Histogram_ResetHint(h.CounterResetHint)
+	hist.Timestamp = timestamp
+	return hist
 }
 
```
This is getting `Histogram` type from memory pool.

tonyli233 commented on prompb/io/prometheus/write/v2/codec.go:
```diff
@@ -25,12 +25,19 @@ import (
 
 // ToMetadata return model metadata from timeseries' remote metadata.
-func (m TimeSeries) ToMetadata(symbols []string) metadata.Metadata {
+func (m *TimeSeries) ToMetadata(symbols []string) metadata.Metadata {
+	if m.Metadata == nil {
+		return metadata.Metadata{
+			Type: model.MetricTypeUnknown,
+			Unit: "",
+			Help: "",
+		}
+	}
```
This is handling missing metadata of time series.

tonyli233 commented on storage/remote/queue_manager.go:
```diff
@@ -1619,16 +1620,33 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
 	for nPending, d := range batch {
-		pendingData[nPending].Samples = pendingData[nPending].Samples[:0]
 		if sendExemplars {
+			for _, exemplar := range pendingData[nPending].Exemplars {
+				for _, lbl := range exemplar.Labels {
+					lbl.ReturnToVTPool()
+				}
+				exemplar.ReturnToVTPool()
+			}
```
This part returns the objects to the memory pool before populating new objects.

tonyli233 commented on storage/remote/queue_manager.go:
```diff
@@ -2136,36 +2164,39 @@ func buildTimeSeries(timeSeries []prompb.TimeSeries, filter func(prompb.TimeSeri
+		*pBuf = make([]byte, req.SizeVT())
 	}
-	err := pBuf.Marshal(req)
+	size, err := req.MarshalToVT(*pBuf)
+	req.Metadata = []*prompb.MetricMetadata{}
+	req.Timeseries = []*prompb.TimeSeries{}
+	req.ReturnToVTPool()
```
We need to reset the `req.Metadata` and `req.Timeseries` since the input `timeSeries` and `metadata` will be reused after the function returns. `req.ReturnToVTPool()` will destroy those fields.

bboreham commented:
I downloaded the code (commit 9218e8c677ef6adcffd6303f3556252aae2c4465) and got this test failure:
```
--- FAIL: TestSampleDelivery (1.51s)
    --- FAIL: TestSampleDelivery/prometheus.WriteRequest-samples,_exemplars,_and_histograms (0.50s)
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0
```

I see you mention a number of points at "Tests to fix:", but apparently not this one.

bboreham commented:
Hello from the bug-scrub!  @tonyli233 do you think you will come back to this?

tonyli233 commented:
> Hello from the bug-scrub! @tonyli233 do you think you will come back to this?

Yes, I will come back to this. More tests are failing now than before since new tests were written and are conflicting with this PR. I think it would be more helpful to address the questions that I posted above in the description before coming back to fix the tests and review. Can I get more feedbacks on that?

bboreham commented:
Hello from the bug-scrub again!

> it would be more helpful to address the questions that I posted above in the description

I'm sorry, I could not detect any questions on a quick read.  I used the browser search function for "?".
Please can you clarify what you would like answered.

bboreham commented:
Hello again from the bug-scrub.  Closing due to lack of response.
